### DISCUSSION TITLE
Incorrect overlay when calling a DrawMoleculeWithHighlights several times

### DISCUSSION THREAD

**Discussion by CyberDreamer** [rdkit/rdkit#6476] [Development]
I need to compose a complex SVG for a single molecule. 

I use a function `DrawMoleculeWithHighlights` to do this. 

I call it in a loop with different parameters `highlight_atom_map` and `highlight_bond_map` but the same `mol`. 

And sometimes it works fine (left side on attachment pic). But sometimes there is an incorrect overlay (right side on attachment pic) when one layer is shifted relative to the other. 
<img width="1285" alt="Снимок экрана 2023-06-22 в 02 31 48" src="https://github.com/rdkit/rdkit/assets/15001410/7e47f482-3b85-46ac-8fff-8bcf356475d5">

How do I fix this or choose a different function in general?

**UPD: 
Add minimal code for reproduction**
```
from rdkit import Chem
from collections import defaultdict
from IPython.display import SVG, display
from rdkit.Chem.Draw import rdMolDraw2D

molecule = Chem.MolFromSmiles(
    "O=C(O[C@H]1CCOC1)N[C@@H](Cc2ccccc2)[C@H](O)CN(CC(C)C)S(=O)(=O)c3ccc(N)cc3")

d2d = rdMolDraw2D.MolDraw2DSVG(500, 500)
rgb_color = (0.0, 1, 1)

color = (0.95, 0.45, 0.45)
highlight_atoms_group = [
               [0, 1, 2, 3, 4], 
               [3, 4, 5, 6, 7], 
               [11, 12, 10, 9, 3, 4, 5, 6, 7], 
               [12, 13, 14, 15, 16, 17]
               ]


for atoms in highlight_atoms_group:
    highlight_atom_map = defaultdict(list)
    highlight_bond_map = defaultdict(list)
    highlight_radii = defaultdict(float)

    for atom in atoms:
        highlight_atom_map[atom].append(color)
        highlight_bond_map[atom].append(color)
        highlight_radii[atom] = 0.2

    d2d.DrawMoleculeWithHighlights(molecule,
                                   legend='',
                                   highlight_atom_map=dict(highlight_atom_map),
                                   highlight_bond_map=dict(highlight_bond_map),
                                   highlight_radii=dict(highlight_radii),
                                   highlight_linewidth_multipliers={})

d2d.FinishDrawing()
svg = d2d.GetDrawingText()
display(SVG(svg))
```

**Comment by DavidACosgrove** [comment]
No one will be able to help you with this unless you include some minimal
example code that reproduces the problem.  You should include the version
of the toolkit your are using, as well.  If it isn't the latest version, it
may be that the problem has already been fixed.

 It may be that DrawMoleculeWithHighlights might help you.  It allows
multiple highlighting on the same molecule in one call.  There is an
example of its use here:
https://rdkit.blogspot.com/2020/04/new-drawing-options-in-202003-release.html



On Wed, Jun 21, 2023 at 11:32 PM Александр Михеев ***@***.***>
wrote:

> I need to compose a complex SVG for a single molecule.
>
> I use a function DrawMoleculeWithHighlights to do this.
>
> I call it in a loop with different parameters highlight_atom_map and
> highlight_bond_map but the same mol.
>
> And sometimes it works fine (left side on attachment pic). But sometimes
> there is an incorrect overlay (right side on attachment pic) when one layer
> is shifted relative to the other.
> [image: Снимок экрана 2023-06-22 в 02 31 48]
> <https://user-images.githubusercontent.com/15001410/247778136-7e47f482-3b85-46ac-8fff-8bcf356475d5.png>
>
> How do I fix this or choose a different function in general?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/rdkit/rdkit/discussions/6476>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACGF2FXT6LV3L5LCQBQZA4LXMNZAFANCNFSM6AAAAAAZPMBGFE>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 
David Cosgrove
Freelance computational chemistry and chemoinformatics developer
http://cozchemix.co.uk

**Reply by CyberDreamer** [reply]
Yes, of course, I added the code. Now I solved the problem by copying the necessary elements, but I'd like a more native solution right from RDkit

**Reply by pippo1990** [reply]
my attempt:

```

from rdkit import Chem
from collections import defaultdict

from IPython.display import SVG, display

from rdkit.Chem.Draw import rdMolDraw2D

from cairosvg import svg2png

from PIL import Image

from io import BytesIO

from copy import deepcopy

molecule = Chem.MolFromSmiles(
    "O=C(O[C@H]1CCOC1)N[C@@H](Cc2ccccc2)[C@H](O)CN(CC(C)C)S(=O)(=O)c3ccc(N)cc3")

d2d = rdMolDraw2D.MolDraw2DSVG(500, 500)

dopts = d2d.drawOptions()

dopts.addAtomIndices = True

dopts.addBondIndices = True


# rgb_color = (0.0, 1, 1)

colors = iter([(0, 0, 1, ), (1, 0, 0, ), (1, 1, 0, ), (0, 1, 0, )])




highlight_atoms_group = [
               [0, 1, 2, 3, 4], 
               [3, 4, 5, 6, 7], 
               [11, 12, 10, 9, 3, 4, 5, 6, 7], 
               [12, 13, 14, 15, 16, 17]
               ]



highlight_bond = {}

for num , atoms in enumerate(highlight_atoms_group):
    
    highlight_bond[num] = []
    
    
print('\n\nhighlight_bond : ', highlight_bond)
    

for number, atoms in enumerate(highlight_atoms_group):
    
    for i in range(molecule.GetNumBonds()):
        
        bond = molecule.GetBondWithIdx(i)
        
        begin_aid = bond.GetBeginAtomIdx()
        
        end_aid = bond.GetEndAtomIdx()
        
        if begin_aid in atoms and end_aid in atoms:
            
            highlight_bond[number].append(i)
            
print('\n\nhighlight_bond : ', highlight_bond)

highlight_atom_map = {}
highlight_bond_map = {}
highlight_radii = {}


colors_copy = deepcopy(colors)

for atoms in highlight_atoms_group:
    
    color = next(colors_copy)
    
    for atom in atoms: 
        
        highlight_atom_map[atom] = [color]
        # highlight_bond_map[atom] = [color]
        highlight_radii[atom] = 0.4
        

colors_copy = deepcopy(colors)

for bonds in highlight_bond.values():
    
    color = next(colors_copy)
    
    for bond in bonds:
        
        # print(bond)
        
        highlight_bond_map[bond] = [color]
        
        
    
    print('+++++++++')
    
    
print('\n\nhighlight_atom_map : ', highlight_atom_map)   

print('\n\nhighlight_bond_map : ', highlight_bond_map)   
    


d2d.DrawMoleculeWithHighlights(molecule,
                                   "try",
                                   highlight_atom_map=highlight_atom_map,
                                   highlight_bond_map=highlight_bond_map,
                                   highlight_radii=highlight_radii,
                                   highlight_linewidth_multipliers={})

d2d.FinishDrawing()
svg = d2d.GetDrawingText()


svg_get = d2d.GetDrawingText()

svg = SVG(svg_get)  # for jupiter ???


print('\n\n\n svg : ' , svg)

svg_data = BytesIO(svg.data.encode())
    
with open('output_2.svg', 'w') as f:
    f.write(svg.data)
        
# png = svg2png(bytestring=svg_data.read()) OR BELOW:

png = svg2png(bytestring=svg_get)

pil_img = Image.open(BytesIO(png)).convert('RGBA')

pil_img.save('pil.png')
    
print(type(pil_img))  

pil_img.show()

```

Not sure that bond highlight make sense as such

![tmp99v3q3wj](https://github.com/user-attachments/assets/58dacff4-0782-4cbb-886a-3b1738695b12)
